aws/resource_aws_backup_plan.go

```diff
@@ -0,0 +1,311 @@
+						"start_window": {
+							Type:     schema.TypeInt,
+							Optional: true,
+							Default:  60,
+						},
+						"completion_window": {
+							Type:     schema.TypeInt,
+							Optional: true,
+							Default:  180,
+						},
+						"lifecycle": {
+							Type:     schema.TypeMap,
+							Optional: true,
+							Elem: &schema.Resource{
+								Schema: map[string]*schema.Schema{
+									"cold_storage_after": {
+										Type:     schema.TypeInt,
+										Optional: true,
+									},
+									"delete_after": {
+										Type:     schema.TypeInt,
+										Optional: true,
+									},
+								},
+							},
+						},
+						"recovery_point_tags": {
+							Type:     schema.TypeMap,
+							Optional: true,
+							Elem:     &schema.Schema{Type: schema.TypeString},
+						},
+					},
+				},
+				Set: resourceAwsPlanRuleHash,
+			},
+			"tags": {
+				Type:     schema.TypeMap,
+				Optional: true,
+				ForceNew: true,
+				Elem:     &schema.Schema{Type: schema.TypeString},
+			},
+			"arn": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"version": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+		},
+	}
+}
+
+func resourceAwsBackupPlanCreate(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).backupconn
+
+	plan := &backup.PlanInput{
+		BackupPlanName: aws.String(d.Get("name").(string)),
+	}
+
+	rules := gatherPlanRules(d)
+
+	plan.Rules = rules
+
+	input := &backup.CreateBackupPlanInput{
+		BackupPlan: plan,
+	}
+
+	if v, ok := d.GetOk("tags"); ok {
+		input.BackupPlanTags = v.(map[string]*string)
+	}
+
+	resp, err := conn.CreateBackupPlan(input)
+	if err != nil {
+		return err
+	}
+
+	d.SetId(*resp.BackupPlanId)
+
+	return resourceAwsBackupPlanRead(d, meta)
+}
+
+func resourceAwsBackupPlanRead(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).backupconn
+
+	input := &backup.GetBackupPlanInput{
+		BackupPlanId: aws.String(d.Id()),
+	}
+
+	resp, err := conn.GetBackupPlan(input)
+	if err != nil {
+		return err
+	}
+
+	rule := &schema.Set{F: resourceAwsPlanRuleHash}
+
+	for _, r := range resp.BackupPlan.Rules {
+		m := make(map[string]interface{})
+
+		if r.CompletionWindowMinutes != nil {
+			m["completion_window"] = *r.CompletionWindowMinutes
+		}
+		if r.Lifecycle != nil {
+			l := map[string]int64{}
+			if r.Lifecycle.DeleteAfterDays != nil {
+				l["delete_after"] = *r.Lifecycle.DeleteAfterDays
+			}
+			if r.Lifecycle.MoveToColdStorageAfterDays != nil {
+				l["cold_storage_after"] = *r.Lifecycle.MoveToColdStorageAfterDays
+			}
+			m["lifecycle"] = l
+		}
+		if r.RecoveryPointTags != nil {
+			m["recovery_point_tags"] = r.RecoveryPointTags
+		}
+		m["rule_name"] = *r.RuleName
+		if r.ScheduleExpression != nil {
+			m["schedule"] = *r.ScheduleExpression
+		}
+		if r.StartWindowMinutes != nil {
+			m["start_window"] = *r.StartWindowMinutes
+		}
+		m["target_vault_name"] = *r.TargetBackupVaultName
+
+		rule.Add(m)
+	}
+	d.Set("rule", rule)
```

Comment:
When using `d.Set()` with aggregate types (TypeList, TypeSet, TypeMap), [we should perform error checking to prevent issues where the code is not properly able to set the Terraform state](https://www.terraform.io/docs/extend/best-practices/detecting-drift.html#error-checking-aggregate-types). e.g.

```suggestion
	if err := d.Set("rule", rule); err != nil {
		return fmt.Errorf("error setting rule: %s", err)
	}
```